File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        HashMap<Integer,Integer> ans=new HashMap<>();
        int total=0;
        int counter=0;
        ans.put(0,1);// for storing 0 as sum so we can consider values from 0 index
        for(int i=0;i<nums.length;i++)
        {
            total+=nums[i];
            if(ans.containsKey(total-k))
            {
                counter+=ans.get(total-k);
            }
           ans.put(total,ans.getOrDefault(total,0)+1);
        }
        return counter;
    }
}

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int,int> mp(nums.size());
        mp[0] = 1;
        int sum = 0;
        int ans = 0;
        for(int i = 0; i<nums.size();i++)
        {
            sum+=nums[i];
            int key = sum -k;
            if(mp.find(key) !=mp.end())
            {
                ans+=mp.find(key)->second;
            }
            mp[sum]++;
        }
        return ans;

    }
};